File: mobile/lib/services/pedidos_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';
import 'api_client.dart';

class PedidosService {
  static final PedidosService instance = PedidosService._internal();
  factory PedidosService() => instance;
  PedidosService._internal();

  Future<List<Pedido>> listarMisPedidos() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/pedidos'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    final data = jsonDecode(response.body) as List;
    return data.map((e) => Pedido.fromJson(e)).toList();
  }

  Future<Pedido> crearDesdeCarrito(int sucursalId) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/pedidos'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({'sucursal_id': sucursalId}),
    );
    _verificar(response);
    return Pedido.fromJson(jsonDecode(response.body));
  }

  Future<Pedido> cancelar(int pedidoId) async {
    final response = await http.put(
      Uri.parse('${ApiClient.baseUrl}/pedidos/$pedidoId/cancelar'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    return Pedido.fromJson(jsonDecode(response.body));
  }

  Future<void> pagar({
    required int pedidoId,
    required String metodo,
    required double monto,
    String? referencia,
  }) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/pagos'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({
        'pedido_id': pedidoId,
        'metodo': metodo,
        'monto': monto,
        'referencia': referencia,
      }),
    );
    _verificar(response);
  }

  // ----- Administración -----
  Future<List<Pedido>> listarTodos({String? estado}) async {
    final query = estado != null ? '?estado=$estado' : '';
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/pedidos/admin/todos$query'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    final data = jsonDecode(response.body) as List;
    return data.map((e) => Pedido.fromJson(e)).toList();
  }

  Future<Pedido> cambiarEstado(int pedidoId, String nuevoEstado) async {
    final response = await http.put(
      Uri.parse('${ApiClient.baseUrl}/pedidos/$pedidoId/estado'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({'nuevo_estado': nuevoEstado}),
    );
    _verificar(response);
    return Pedido.fromJson(jsonDecode(response.body));
  }

  Future<Pedido> crearVentaPresencial({
    required int usuarioId,
    required int sucursalId,
    required String metodoPago,
    required List<Map<String, int>> items,
  }) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/pedidos/venta-presencial'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({
        'usuario_id': usuarioId,
        'sucursal_id': sucursalId,
        'metodo_pago': metodoPago,
        'items': items,
      }),
    );
    _verificar(response);
    return Pedido.fromJson(jsonDecode(response.body));
  }

  void _verificar(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw Exception(_extraerError(response.body));
    }
  }

  String _extraerError(String body) {
    try {
      final data = jsonDecode(body);
      if (data is Map && data['detail'] != null) return data['detail'].toString();
    } catch (_) {}
    return 'No se pudo completar la operación del pedido';
  }
}

File: frontend/src/app/core/services/pedidos.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface DetallePedido {
  id: number;
  pedido_id: number;
  variante_id: number;
  cantidad: number;
  precio_unitario: number;
  subtotal: number;
}

export interface Pedido {
  id: number;
  usuario_id: number;
  fecha_pedido: string;
  estado: string;
  total: number;
  detalles: DetallePedido[];
}

export interface ItemVentaPresencial {
  variante_id: number;
  cantidad: number;
}

@Injectable({
  providedIn: 'root'
})
export class PedidosService {
  private http = inject(HttpClient);
  private apiUrl = 'https://menstyle-api-n77g.onrender.com/api/pedidos';

  private headers(): HttpHeaders {
    return new HttpHeaders({
      Authorization: `Bearer ${localStorage.getItem('access_token')}`
    });
  }

  listarMisPedidos(estado?: string): Observable<Pedido[]> {
    const query = estado ? `?estado=${estado}` : '';
    return this.http.get<Pedido[]>(
      `${this.apiUrl}${query}`,
      { headers: this.headers() }
    );
  }

  obtener(id: number): Observable<Pedido> {
    return this.http.get<Pedido>(
      `${this.apiUrl}/${id}`,
      { headers: this.headers() }
    );
  }

  crearDesdeCarrito(sucursalId: number): Observable<Pedido> {
    return this.http.post<Pedido>(
      this.apiUrl,
      { sucursal_id: sucursalId },
      { headers: this.headers() }
    );
  }

  cancelar(id: number): Observable<Pedido> {
    return this.http.put<Pedido>(
      `${this.apiUrl}/${id}/cancelar`,
      {},
      { headers: this.headers() }
    );
  }

  // ----- Administración -----
  listarTodos(estado?: string): Observable<Pedido[]> {
    const query = estado ? `?estado=${estado}` : '';
    return this.http.get<Pedido[]>(
      `${this.apiUrl}/admin/todos${query}`,
      { headers: this.headers() }
    );
  }

  cambiarEstado(id: number, nuevoEstado: string, sucursalId?: number): Observable<Pedido> {
    return this.http.put<Pedido>(
      `${this.apiUrl}/${id}/estado`,
      { nuevo_estado: nuevoEstado, sucursal_id: sucursalId ?? null },
      { headers: this.headers() }
    );
  }

  crearVentaPresencial(
    usuarioId: number,
    sucursalId: number,
    metodoPago: string,
    items: ItemVentaPresencial[]
  ): Observable<Pedido> {
    return this.http.post<Pedido>(
      `${this.apiUrl}/venta-presencial`,
      { usuario_id: usuarioId, sucursal_id: sucursalId, metodo_pago: metodoPago, items },
      { headers: this.headers() }
    );
  }
}
